### PR TITLE
build: Automatically reduce code

### PR DESCRIPTION
**Pull request by rbradford**
Automatically reduce code that is pushed to the topic/virt-x86 branch and push
that to a new branch. That will trigger a build for that branch.

Still experimental so that branch is called experiment/automatic-removal

Signed-off-by: Rob Bradford <robert.bradford@intel.com>

### CODE REVIEW

**Comment by rbradford** [comment]
As this is designed to only work on specific branches e.g. topic/virt-x86 i manually tested this by updating the pipleline via the UI.

File: Jenkinsfile [hunk 1]
```diff
@@ -1,46 +1,105 @@
...
+			}
+			stage ('Remove unused files') {
+				sh "cat $WORKSPACE/used-aarch64.txt $WORKSPACE/used-x86-64.txt | sort | uniq > $WORKSPACE/used-c-files.txt"
+				sh "find . | grep \\\\.[ch]\$ | sed 's#^./##' | sort | uniq > $WORKSPACE/all-c-files.txt"
+				sh "comm -23 $WORKSPACE/all-c-files.txt $WORKSPACE/used-c-files.txt | tee $WORKSPACE/unused-files.txt"
+				sh "wc -l *.txt"
+				sh "git branch -D tmp-branch || true"
+				sh "git checkout -b tmp-branch"
+				sh 'cat $WORKSPACE/unused-files.txt | grep -v ^linux-headers | xargs -I {} -n 1 sh -c "git rm {} || true"'
+				sh 'git commit -m "Remove unused files" --author "NEMU CI <noreply@intel.com>" -a'
```
Comment (sameo) [inline Jenkinsfile]: Could we please also add stats about the amount of code that got removed? something along the lines of what our `cloc-change.sh` currently generates.

File: Jenkinsfile [hunk 2]
```diff
@@ -1,46 +1,105 @@
...
+			stage ('Download results') {
+				azureDownload storageCredentialId: 'nemu-jenkins-storage-account', 
+						downloadType: 'container',
+						containerName: env.BUILD_TAG.replaceAll("%2F","-")
+			}
+			stage ('Remove unused files') {
+				sh "cat $WORKSPACE/used-aarch64.txt $WORKSPACE/used-x86-64.txt | sort | uniq > $WORKSPACE/used-c-files.txt"
+				sh "find . | grep \\\\.[ch]\$ | sed 's#^./##' | sort | uniq > $WORKSPACE/all-c-files.txt"
+				sh "comm -23 $WORKSPACE/all-c-files.txt $WORKSPACE/used-c-files.txt | tee $WORKSPACE/unused-files.txt"
+				sh "wc -l *.txt"
```
Comment (sameo) [inline Jenkinsfile]: What is this for?

**Review by sameo — COMMENTED** [review]
Just a couple of comments. The logic and the idea of pushing the reduced code base to an experimental branch seems fine to me for now.

File: Jenkinsfile (hunk 2 above)
Comment (rbradford) [reply Jenkinsfile]: It's a poor man's CLOC change. I'll switch to cloc-change.sh

File: Jenkinsfile (hunk 1 above)
Comment (rbradford) [reply Jenkinsfile]: Do you want cloc-change.sh relative to the previous commit on that branch or relative to topic/virt-x86. I feel the latter is more useful.

Comment (sameo) [reply Jenkinsfile]: The latter, definitely.

**Comment by rbradford** [comment]
All done @sameo

**Review by sameo — APPROVED** [review]
LGTM

**Pull request merged**